tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py

```diff
@@ -329,18 +335,32 @@ def port_forward_pod(
                         raise PortForwardingError(
                             'Error forwarding port, kubectl return '
                             f'code {return_code}, output {errors}')
-                elif output != expected:
-                    raise PortForwardingError(
-                        f'Error forwarding port, unexpected output {output}')
+                    # If there is no output, and the subprocess is not exiting,
+                    # continue waiting for the log line.
+                    continue
+
+                # Validate output log
+                if local_port:
+                    if output != local_port_expected:
+                        raise PortForwardingError(
+                            f'Error forwarding port, unexpected output {output}'
+                        )
                 else:
-                    logger.info(output)
-                    break
+                    groups = local_port_re.search(output)
+                    if groups is None:
+                        raise PortForwardingError(
+                            f'Error forwarding port, unexpected output {output}'
+                        )
+                    local_port = int(groups[1])
+
+                logger.info(output)
+                break
         except Exception:
             self.port_forward_stop(pf)
             raise
 
         # TODO(sergiitk): return new PortForwarder object
-        return pf
+        return local_port, pf
```

Comment:
Nit: maybe we could create a `PortForwarder` class that will wrap `subprocess.Popen` and the port?

Reply:
We also could move `k8s_namespace.port_forward_stop.port_forward_stop(pf):` static method to a concrete method of `PortForwarder`.